src/v0/destinations/klaviyo/batchUtil.js

```diff
@@ -0,0 +1,202 @@
+  isDefinedAndNotNull,
+} = require('../../util');
+const { JSON_MIME_TYPE } = require('../../util/constant');
+const { BASE_ENDPOINT, CONFIG_CATEGORIES, MAX_BATCH_SIZE, revision } = require('./config');
+const { buildRequest, getSubscriptionPayload } = require('./util');
+
+/**
+ * This function groups the subscription responses on list id
+ * @param {*} subscribeResponseList
+ * @returns
+ * Example subsribeResponseList =
+ * [
+ * { payload: {id:'list_id', profile: {}}, metadata:{} },
+ * { payload: {id:'list_id', profile: {}}, metadata:{} }
+ * ]
+ */
+const groupSubscribeResponsesUsingListIdV2 = (subscribeResponseList) => {
+  const subscribeEventGroups = lodash.groupBy(
+    subscribeResponseList,
+    (event) => event.payload.listId,
+  );
+  return subscribeEventGroups;
+};
+
+/**
+ * This function takes susbscription as input and batches them into a single request body
+ * @param {subscription}
+ * subscription= {listId, subscriptionProfileList}
+ */
+const generateBatchedSubscriptionRequest = (subscription, destination) => {
+  const batchEventResponse = defaultBatchRequestConfig();
+  // if( !isDefinedAndNotNull(subscription) )
+  // fetching listId from first event as listId is same for all the events
+  const profiles = []; // list of profiles to be subscribed
+  const { listId, subscriptionProfileList } = subscription;
+  subscriptionProfileList.forEach((profileList) => profiles.push(...profileList));
+  batchEventResponse.batchedRequest.body.JSON = getSubscriptionPayload(listId, profiles);
+  batchEventResponse.batchedRequest.endpoint = `${BASE_ENDPOINT}/api/profile-subscription-bulk-create-jobs`;
+  batchEventResponse.batchedRequest.headers = {
+    Authorization: `Klaviyo-API-Key ${destination.Config.privateApiKey}`,
+    'Content-Type': JSON_MIME_TYPE,
+    Accept: JSON_MIME_TYPE,
+    revision,
+  };
+  return batchEventResponse.batchedRequest;
+};
+
+/**
+ * This function updates batchedRequest with profile requests
+ * @param {*} profiles
+ * @param {*} batchedRequest
+ * @param {*} destination
+ */
+const updateBatchEventResponseWithProfileRequests = (profiles, batchedRequest, destination) => {
+  const profilesRequests = [];
+  profiles.forEach((profile) => {
+    profilesRequests.push(buildRequest(profile, destination, CONFIG_CATEGORIES.IDENTIFYV2));
+  });
+  // we are keeping profiles request prior to subscription ones as first profile creation and then subscription should happen
+  batchedRequest.unshift(...profilesRequests);
+};
+
+/**
+ * this function populates profileSubscriptionAndMetadataArr with respective profiles based upon common metadata
+ * @param {*} profileSubscriptionAndMetadataArr
+ * @param {*} metadataIndexMap
+ * @param {*} profiles
+ * @returns updated profileSubscriptionAndMetadataArr obj
+ */
+const populateArrWithRespectiveProfileData = (
+  profileSubscriptionAndMetadataArr,
+  metadataIndexMap,
+  profiles,
+) => {
+  const updatedPSMArr = profileSubscriptionAndMetadataArr;
```

Comment:
If you want to get updatePSMarr use deepClone 

Ref: https://stackoverflow.com/questions/597588/how-do-you-clone-an-array-of-objects-in-javascript
```
example
array2 = structuredClone(array1);
```